### ISSUE TITLE
32-bit test failure: test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed

### ISSUE THREAD

**Issue by error10** [bitcoin/bitcoin#14580]
A test in the bitcoin test suite is failing for me while running the sha256d64 test cases in crypto_tests.cpp, but only on 32-bit distro. It passes on 64-bit.

I expected the test suite to pass.

The test suite log is:

```
$ cat test/crypto_tests.cpp.log
Running 11 test cases...
Entering test module "Bitcoin Test Suite"
test/crypto_tests.cpp(23): Entering test suite "crypto_tests"
test/crypto_tests.cpp(217): Entering test case "ripemd160_testvectors"
test/crypto_tests.cpp(217): Leaving test case "ripemd160_testvectors"; testing time: 343157us
test/crypto_tests.cpp(233): Entering test case "sha1_testvectors"
test/crypto_tests.cpp(233): Leaving test case "sha1_testvectors"; testing time: 200719us
test/crypto_tests.cpp(249): Entering test case "sha256_testvectors"
test/crypto_tests.cpp(249): Leaving test case "sha256_testvectors"; testing time: 518373us
test/crypto_tests.cpp(271): Entering test case "sha512_testvectors"
test/crypto_tests.cpp(271): Leaving test case "sha512_testvectors"; testing time: 1071278us
test/crypto_tests.cpp(308): Entering test case "hmac_sha256_testvectors"
test/crypto_tests.cpp(308): Leaving test case "hmac_sha256_testvectors"; testing time: 10312us
test/crypto_tests.cpp(361): Entering test case "hmac_sha512_testvectors"
test/crypto_tests.cpp(361): Leaving test case "hmac_sha512_testvectors"; testing time: 22518us
test/crypto_tests.cpp(429): Entering test case "aes_testvectors"
test/crypto_tests.cpp(429): Leaving test case "aes_testvectors"; testing time: 3358us
test/crypto_tests.cpp(445): Entering test case "aes_cbc_testvectors"
test/crypto_tests.cpp(445): Leaving test case "aes_cbc_testvectors"; testing time: 4206us
test/crypto_tests.cpp(497): Entering test case "chacha20_testvector"
test/crypto_tests.cpp(497): Leaving test case "chacha20_testvector"; testing time: 3051us
test/crypto_tests.cpp(527): Entering test case "countbits_tests"
test/crypto_tests.cpp(527): Leaving test case "countbits_tests"; testing time: 63452us
test/crypto_tests.cpp(549): Entering test case "sha256d64"
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(549): Leaving test case "sha256d64"; testing time: 5115us
test/crypto_tests.cpp(23): Leaving test suite "crypto_tests"; testing time: 2246048us
Leaving test module "Bitcoin Test Suite"; testing time: 2246167us

*** 32 failures are detected in the test module "Bitcoin Test Suite"
```

This is occurring on Fedora 29 i686 while building bitcoin. The builder has 4GB RAM. This environment includes boost 1.66.0 and openssl 1.1.1. The test passes while building bitcoin on Fedora 29 x86_64 on the same machine. The test also passed on Fedora 28 i686, which has boost 1.66.0 and openssl 1.1.0i.

This is bitcoin 0.17.0 obtained from thhe github release tagged v0.17.0.

I glanced at the test but I didn't see anything obvious that might cause it to fail on a 32-bit system. I would appreciate a hint as to where to look next.

**Comment by MarcoFalke** [comment]
Could you also check if branch 16 and branch master are affected, please?

**Comment by error10** [comment]
> Could you also check if branch 16 and branch master are affected, please?

The test passes on v0.16.3, and fails identically as above on master.

**Comment by error10** [comment]
> Do you have the same issue with release binaries?

Good call. I downloaded bitcoin-0.17.0-i686-pc-linux-gnu.tar.gz, and its included bin/test_bitcoin passes. So, I guess I don't have the same issue with release binaries.

**Comment by MarcoFalke** [comment]
I believe this also happens 32bit Ubuntu Cosmic.

**Comment by error10** [comment]
> I believe this also happens 32bit Ubuntu Cosmic.

I don't just believe. I tested it, and sure enough it blows up on Ubuntu 18.10 i386, identically as above.

Cosmic also has openssl 1.1.1, so perhaps it's something specific to 1.1.1? A 32-bit openssl bug maybe?

**Comment by error10** [comment]
> I am running 0.15.1 that won't upgrade on a raspberry pi 3 B. If I can help in any way with testing I would be glad to.

You can just run a 64-bit distro on the RPi 3. Until Raspbian goes 64-bit, Fedora is probably your best bet. And then you can just pick up my aarch64 repo if you want.

**Comment by develop7** [comment]
the 0.17.1 is affected too, see https://build.opensuse.org/package/live_build_log/home:develop7:branches:network:cryptocurrencies/bitcoin/openSUSE_Tumbleweed/i586

**Comment by MarcoFalke** [comment]
As a workaround you could build from our `./depends` (which comes with an older openssl)

**Comment by error10** [comment]
If 32-bit support were officially just dropped, I wouldn't complain. Aside from ancient Raspberry Pi 1 and 2s and the occasional antique netbook, there's virtually no hardware out there these days that requires a 32-bit OS. And the resource requirements of Bitcoin seem unlikely to shrink in the future.

**Comment by MarcoFalke** [comment]
Oh, we didn't drop support for 32-bit. This is just a bug introduced in a newer version of openssl, but no one looked into it to understand what happened or how it could be worked around.

**Comment by MarcoFalke** [comment]
I am seeing this running the gitian build (which is built with depends) on ubuntu_xenial_armhf ... :weary:

**Comment by MarcoFalke** [comment]
Steps to reproduce on fresh Ubuntu Bionic 64-bit:


```
dpkg --add-architecture i386
apt update && apt install wget libc6:i386 libncurses5:i386 libstdc++6:i386
wget https://bitcoincore.org/bin/bitcoin-core-0.18.0/bitcoin-0.18.0-i686-pc-linux-gnu.tar.gz
tar -xzf bitcoin-0.18.0-i686-pc-linux-gnu.tar.gz
./bitcoin-0.18.0/bin/test_bitcoin -t crypto_tests
Running 11 test cases...
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed

*** 32 failures are detected in the test module "Bitcoin Test Suite"

**Comment by MarcoFalke** [comment]
Why is the travis 32-bit not failing?

https://travis-ci.org/bitcoin/bitcoin/jobs/527334881#L2853

**Comment by MarcoFalke** [comment]
So I guess it has nothing to do with openssl

**Comment by MarcoFalke** [comment]
I can reproduce also on vanilla fedora 30 - 32 bit


![Schermafdruk van 2019-05-02 17-09-39](https://user-images.githubusercontent.com/6399679/57107143-37c9ab80-6cfd-11e9-9ac1-dcb922a8d6a2.png)

**Comment by sipa** [comment]
@MarcoFalke What does the "Using the '%s' SHA256 implementation" say on that system/build? Also what does `cat /proc/cpuinfo | fgrep flags` say?

**Comment by MarcoFalke** [comment]
```
$ ./src/bitcoind -printtoconsole -regtest | grep SHA
2019-05-02T21:19:53Z Using the 'standard,sse41(4way),avx2(8way)' SHA256 implementation


$ cat /proc/cpuinfo | fgrep flags | uniq -c
      8 flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt xsaveopt xsavec xgetbv1 xsaves arat umip

**Comment by sipa** [comment]
I believe this is a GCC bug, filed as https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90348

**Comment by practicalswift** [comment]
@sipa Excellent work! This sure looks like a miscompiler bug in gcc: the test case does not contain any UB that would give gcc the leeway to behave this way.

**Comment by practicalswift** [comment]
Can be closed now that #15983 has been merged?

**Comment by hebasto** [comment]
> I believe this is a GCC bug, filed as https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90348

It looks like the bug has been [fixed](https://github.com/gcc-mirror/gcc/commit/551aa75778a4c5165d9533cd447c8fc822f583e1) in GCC 12.1+.
